test/functorch/test_control_flow.py

```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Only init given
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        result_init = scan_fct(op, input=[], dim=dim, reverse=reverse, init=init)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(result_init, result_exp)
+        self.assertEqual(result_init, init)
+
+        x = torch.randn(3, 1, 2, device=device)
+        init = torch.randn(3, 1, 2, device=device)
+        dim = 1
+
+        op, op_pt = (add, torch.cumsum)
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        # Init scalar
+        init = 1.0
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong shape (More than one time step)
+        init = torch._ops.ops.aten.slice(x, dim, 0, 2, 1)
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong shape (Other dim different)
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        init = torch.tile(init, (1, 2, 1))
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong pytree
+        init = (
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+        )
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Correct case
+        op, op_pt = (add, torch.cumsum)
+        x = torch.randn(3, 2, 2, device=device)
+        dim = 1
+
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+        result_exp = _fake_scan(op, x, dim=dim, reverse=reverse)
+
+        if reverse:
+            x = torch.flip(x, [dim])
+
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        self.assertEqual(result, result_exp)
+        self.assertEqual(result_init, result_exp)
+        if not reverse:
+            result_exp_PT = op_pt(x, dim)
+            self.assertEqual(result, result_exp_PT)
+            self.assertEqual(result_init, result_exp_PT)
+
+    @unittest.skipIf(not SM70OrLater, "triton")
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_compile_cnt(self, reverse, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
```

Comment:
Do we only support combine functions that takes two tensor inputs? Can the combine_fn support pytrees?
```
def combine(x_dict, y_dict):
  ...
```

Reply:
No, like in the case of `associative_scan`, the combine_fn supports pytrees as inputs. I have added testcases for that.